docs/sdks/c-sharp/index-v2.md

```diff
@@ -0,0 +1,860 @@
+
+### Callback `OnConnectError`
+
+```csharp
+class DbConnectionBuilder<DbConnection>
+{
+    public DbConnectionBuilder<DbConnection> OnConnectError(Action<ErrorContext, SpacetimeDbException> callback);
+}
+```
+
+Chain a call to `.OnConnectError(callback)` to your builder to register a callback to run when your connection fails.
+
+A known bug in the SpacetimeDB Rust client SDK currently causes this callback never to be invoked. [`OnDisconnect`](#callback-ondisconnect) callbacks are invoked instead.
+
+### Callback `OnDisconnect`
+
+```csharp
+class DbConnectionBuilder<DbConnection>
+{
+    public DbConnectionBuilder<DbConnection> OnDisconnect(Action<ErrorContext, SpacetimeDbException> callback);
+}
+```
+
+Chain a call to `.OnDisconnect(callback)` to your builder to register a callback to run when your `DbConnection` disconnects from the remote module, either as a result of a call to [`Disconnect`](#method-disconnect) or due to an error.
+
+### Method `WithToken`
+
+```csharp
+class DbConnectionBuilder<DbConnection>
+{
+    public DbConnectionBuilder<DbConnection> WithToken(string token = null);
+}
+```
+
+Chain a call to `.WithToken(token)` to your builder to provide an OpenID Connect compliant JSON Web Token to authenticate with, or to explicitly select an anonymous connection. If this method is not called or `None` is passed, SpacetimeDB will generate a new `Identity` and sign a new private access token for the connection.
+
+### Method `Build`
+
+```csharp
+class DbConnectionBuilder<DbConnection>
+{
+    public DbConnection Build();
+}
+```
+
+After configuring the connection and registering callbacks, attempt to open the connection.
+
+## Advance the connection and process messages
+
+In the interest of supporting a wide variety of client applications with different execution strategies, the SpacetimeDB SDK allows you to choose when the `DbConnection` spends compute time and processes messages. If you do not arrange for the connection to advance by calling one of these methods, the `DbConnection` will never advance, and no callbacks will ever be invoked.
+
+| Name                                        | Description                                           |
+|---------------------------------------------|-------------------------------------------------------|
+| [`FrameTick` method](#method-frametick)     | Process messages on the main thread without blocking. |
+
+#### Method `FrameTick`
+
+```csharp
+class DbConnection {
+    public void FrameTick();
+}
+```
+
+`FrameTick` will advance the connection until no work remains or until it is disconnected, then return rather than blocking. Games might arrange for this message to be called every frame.
```

Comment:
```suggestion
`FrameTick` will advance the connection until no work remains or until it is disconnected, then return rather than blocking. Games might arrange for this message to be called every frame.

It is not advised to run `FrameTick` on a background thread, since it modifies [`dbConnection.Db`](#property-db). If main thread code is also accessing the `Db`, it may observe data races when `FrameTick` runs on another thread.

(Note that the SDK already does most of the work for parsing messages on a background thread. `FrameTick()` does the minimal amount of work needed to apply updates to the `Db`.)
```